cpp/src/arrow/dataset/dataset_internal.h

```diff
@@ -0,0 +1,197 @@
+
+  Status SetInputSchema(std::shared_ptr<Schema> from) {
+    from_ = std::move(from);
+
+    for (int i = 0; i < to_->num_fields(); ++i) {
+      const auto& field = to_->field(i);
+      int matching_index = from_->GetFieldIndex(field->name());
+
+      if (matching_index != kNoMatch) {
+        if (!from_->field(matching_index)->Equals(field)) {
+          return Status::Invalid("fields had matching names but were not equivalent ",
+                                 from_->field(matching_index)->ToString(), " vs ",
+                                 field->ToString());
+        }
+
+        // Mark column i as not missing by setting missing_columns_[i] to nullptr
+        missing_columns_[i] = nullptr;
+      } else {
+        // Mark column i as missing by setting missing_columns_[i]
+        // to a non-null placeholder.
+        RETURN_NOT_OK(
+            MakeArrayOfNull(pool_, to_->field(i)->type(), 0, &missing_columns_[i]));
+      }
+
+      column_indices_[i] = matching_index;
+    }
+    return Status::OK();
+  }
+
+ private:
+  static constexpr int kNoMatch = -1;
+
+  Status ResizeMissingColumns(int64_t new_length) {
+    // TODO(bkietz) MakeArrayOfNull could use fewer buffers by reusing a single zeroed
+    // buffer for every buffer in every column which is null
+    for (int i = 0; i < to_->num_fields(); ++i) {
+      if (missing_columns_[i] == nullptr) {
+        continue;
+      }
+      if (scalars_[i] == nullptr) {
+        RETURN_NOT_OK(MakeArrayOfNull(pool_, missing_columns_[i]->type(), new_length,
+                                      &missing_columns_[i]));
+        continue;
+      }
+      RETURN_NOT_OK(
+          MakeArrayFromScalar(pool_, *scalars_[i], new_length, &missing_columns_[i]));
+    }
+    missing_columns_length_ = new_length;
+    return Status::OK();
+  }
+
+  MemoryPool* pool_;
+  std::shared_ptr<Schema> from_, to_;
+  int64_t missing_columns_length_ = 0;
+  std::vector<std::shared_ptr<Array>> missing_columns_;
+  std::vector<int> column_indices_;
+  std::vector<std::shared_ptr<Scalar>> scalars_;
+};
+
+constexpr int RecordBatchProjector::kNoMatch;
+
+/// Wraps a RecordBatchIterator and projects each yielded batch using the given projector.
+///
+/// Note that as with RecordBatchProjector, ProjectedRecordBatchReader is most efficient
+/// when projecting record batches with a consistent schema (for example batches from a
+/// table), but it can project record batches having any schema.
+class ProjectedRecordBatchReader : public RecordBatchReader {
+ public:
+  static Status Make(MemoryPool* pool, std::unique_ptr<RecordBatchProjector> projector,
+                     std::unique_ptr<RecordBatchIterator> wrapped,
+                     std::unique_ptr<RecordBatchIterator>* out) {
+    out->reset(
+        new ProjectedRecordBatchReader(pool, std::move(projector), std::move(wrapped)));
+    return Status::OK();
+  }
+
+  Status ReadNext(std::shared_ptr<RecordBatch>* out) override {
```

Comment:
In ARROW-6242, there's `MaybeMapIterator` which you could just wrap the call to `projector->Project` into a lambda.

Reply:
That sounds fine; I'll remove this ProjectedRecordBatchReader then

Reply:
No, keep it, it's still useful, I was just adding a note that you could achieve the same with a tiny refactor (if ARROW-6242 lands).

Reply:
I'll wait to delete until after ARROW-6242, then